### PR TITLE
Ocean: poolpair graph

### PR DESCRIPTION
**Pull request by canonbrother**
## Summary

- `get best path` and `list paths`
- `list swappable tokens`

### CODE REVIEW

**Comment by canonbrother** [comment]
Checked

```
list
    ✕ should list (25 ms)
    ✓ should list with pagination (9 ms)
    ✓ should list with undefined next pagination (3 ms)
  get
    ✕ should get (4 ms)
    ✕ should throw error while getting non-existent poolpair (6 ms)
  get best path
    ✓ should be bidirectional swap path - listPaths(a, b) === listPaths(b, a) (6 ms)
    ✕ should get best swap path - 2 legs (15 ms)
    ✕ should get correct swap path - 3 legs (5 ms)
    ✓ should ignore correct swap path > 3 legs (3 ms)
    ✕ should return direct path even if composite swap paths has greater return (7 ms)
    ✕ should deduct commission fee - 1 leg (3 ms)
    ✕ should deduct commission and dex fees - 2 legs (5 ms)
    ✓ should have no swap path - isolated token H (4 ms)
    ✕ should throw error for invalid tokenId (5 ms)
  get all paths
    ✓ should be bidirectional swap path - listPaths(a, b) === listPaths(b, a) (4 ms)
    ✓ should get correct swap path - 2 legs (2 ms)
    ✓ should get correct swap path - 3 legs (2 ms)
    ✓ should ignore correct swap paths > 3 legs (2 ms)
    ✕ should get multiple swap paths (6 ms)
    ✕ should handle cyclic swap paths (3 ms)
    ✓ should have no swap path - isolated token H (2 ms)
    ✕ should throw error when fromToken === toToken (1 ms)
    ✕ should throw error for invalid tokenId (3 ms)
  get list swappable tokens
    ✓ should list correct swappable tokens (6 ms)
    ✓ should not show status:false tokens (2 ms)
    ✓ should list no tokens for token that is not swappable with any (1 ms)
    ✕ should throw error for invalid / non-existent tokenId (2 ms)
 ```
 
 TODO
 ```
  - list
    - from_with_id
        - name
        - getTotalLiquidityUsd
        - getApr
        - getUSDVolume
        
error msg
Expected: "Unable to find poolpair"
Received: "Pool not found"

error handling
Received promise resolved instead of rejected
Resolved to value: {"at": 1713929344136, "code": 404, "message": "Token not found", "type": "NotFound", "url": "/v0/regtest/poolpairs/paths/best/from/-1/to/1"}

fees cal
-   "estimatedReturn": "0.50000000",
-   "estimatedReturnLessDexFees": "0.50000000",
+   "estimatedReturn": "1.00000000",
+   "estimatedReturnLessDexFees": "1.00000000",

-         "ab": "0.14285714",
+        "ab": "0.14285713",

 -   "estimatedReturnLessDexFees": "6.09336000",
+   "estimatedReturnLessDexFees": "7.00000000",
 ```

File: lib/ain-ocean/src/api/poolpairs_path.rs [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+    }
+
+    Ok(EstimatedLessDexFeeInfo{
+        estimated_return: estimated_return.ceil(),
+        estimated_return_less_dex_fees: estimated_return_less_dex_fees.ceil(),
+    })
+}
+
+pub async fn sync_token_graph(ctx: &Arc<AppContext>) {
+  let mut interval = tokio::time::interval(Duration::from_secs(120));
```
Comment (canonbrother) [inline lib/ain-ocean/src/api/poolpairs_path.rs:323]: js testsuite process longer time due to ensure sync graph is done after 120s

File: lib/ain-ocean/src/api/poolpairs.rs [hunk 2]
```diff
@@ -426,36 +416,152 @@ async fn list_pool_swaps_verbose(
...
+        };
+        token_ids.insert(token_id);
+        let graph = ctx.services.token_graph.lock().clone();
+        let edges = graph.edges(token_id).collect::<Vec<_>>();
+        for edge in edges {
+            token_ids = recur(ctx, token_ids.clone(), edge.0);
+            token_ids = recur(ctx, token_ids.clone(), edge.1);
+        }
+        token_ids
+    }
```
Comment (Jouzo) [inline lib/ain-ocean/src/api/poolpairs.rs:447]: Do we have to both take a mut ref and return the HashSet ?
We can lock only once.
We don't need to clone the set.

Something like :

```
let mut token_ids = HashSet::new();

    fn recur(ctx: &Arc<AppContext>, graph: &UnGraphMap<u32, String>, token_ids: &mut HashSet<u32>, token_id: u32) {
        if token_ids.contains(&token_id) {
            return
        };
        token_ids.insert(token_id);
        let edges = graph.edges(token_id).collect::<Vec<_>>();
        for edge in edges {
            recur(ctx, graph, token_ids, edge.0);
            recur(ctx, graph, token_ids, edge.1);
        }
    }

    {
        let graph = ctx.services.token_graph.lock().clone();
        recur(&ctx, &graph, &mut token_ids, token_id.parse::<u32>()?);
    }

    token_ids.remove(&token_id.parse::<u32>()?);

File: lib/ain-ocean/src/api/poolpairs_path.rs [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+pub async fn get_token_identifier(ctx: &Arc<AppContext>, id: &str) -> Result<TokenIdentifier> {
+    let (id, token) = get_token_cached(ctx, id).await?;
+    Ok(TokenIdentifier{
+        id,
+        name: token.name,
+        symbol: token.symbol.clone(),
+        display_symbol: parse_dat_symbol(&token.symbol),
+    })
+
+}
```
Comment (Jouzo) [inline lib/ain-ocean/src/api/poolpairs_path.rs:112]:
```suggestion
pub async fn get_token_identifier(ctx: &Arc<AppContext>, id: &str) -> Result<TokenIdentifier> {
    let (id, token) = get_token_cached(ctx, id).await?;
    Ok(TokenIdentifier{
        id,
        display_symbol: parse_dat_symbol(&token.symbol),
        name: token.name,
        symbol: token.symbol,
    })
}
```
Removing the clone

**Review by Jouzo — APPROVED** [review]

**Pull request merged**